File: frontend/src/app/models/SalesOrder.ts
import { BaseModel } from './BaseModel';
import { CompanyDto } from './Company';
import { DeliveryOrderDto } from './DeliveryOrder';
import { InvoiceDto } from './Invoice';
import { ProductServiceDto } from './ProductService';
import { ProjectDto } from './Project';
import { PurchaseOrderDto } from './PurchaseOrder';
import { QuotationDto } from './Quotation';
import { UserDto } from './User';

export interface SalesOrderDto extends BaseModel {
  salesOrderNo: string;
  clientId: string;
  client: CompanyDto;
  companyId?: string;
  company?: CompanyDto;
  quotationId?: string;
  quotation?: QuotationDto;
  projectId?: string;
  project?: ProjectDto;
  soDate: Date;
  status: string;
  clientPOAttachment: string;
  clientPODate: Date;
  clientPONumber: string;
  discount: number;
  subTotal: number;
  taxAmount: number;
  totalAmount: number;
  notes?: string;
  paymentTerms?: string;
  remarks?: string;
  salesOrderItems?: SalesOrderItem[];
  salesOrderStatusHistories: SalesOrderStatusHistory[];
  deliveryOrders: DeliveryOrderDto[];
  invoices: InvoiceDto[];
  purchaseOrders: PurchaseOrderDto[];
}

export interface SalesOrderItem extends BaseModel {
  salesOrderId: string;
  salesOrder: SalesOrderDto;
  productServiceId: string;
  productService: ProductServiceDto;
  rowType: string;
  item: string;
  description: string;
  quantity: number;
  quantityDelivered: number;
  quantityAllocated: number;
  quantityRemaining: number;
  unit: string;
  unitPrice: number;
  discount: number;
  taxRate: number;
  totalPrice: number;
  sortOrder: number;
}

export interface SOItemBase {
  id: string;
  salesOrderId: string;
  productServiceId: string;
  item: string;
  rowType: string;
  description: string;
  quantity: number;
  unit: string;
  unitPrice: number;
  discount: number;
  taxRate: number;
  totalPrice: number;
  sortOrder: number;
}

export interface CreateSalesOrderRequest {
  salesOrderNo: string;
  companyId: string;
  clientId: string;
  projectId: string;
  quotationId: string;
  soDate: Date;
  taxAmount: number;
  discount: number;
  subTotal: number;
  totalAmount: number;
  notes: string;
  remarks: string;
  paymentTerms?: string;
  clientPOAttachment?: File;
  clientPODate: Date;
  clientPONumber: string;
  salesOrderItems: SOItemBase[];
}

export interface UpdateSalesOrderRequest {
  id: string;
  salesOrderNo: string;
  companyId: string;
  clientId: string;
  projectId: string;
  quotationId: string;
  paymentTerms?: string;
  soDate: Date;
  subTotal: number;
  taxAmount: number;
  discount: number;
  totalAmount: number;
  notes: string;
  remarks: string;
  clientPOAttachment?: File;
  clientPODate: Date;
  clientPONumber: string;
  salesOrderItems: SOItemBase[];
}

export interface SalesOrderStatusHistory {
  id: string;
  salesOrderId: string;
  salesOrder: SalesOrderDto;
  status: string;
  actionAt: Date;
  actionUser: UserDto;
  remarks: string;
}

export interface UpdateSalesOrderStatusRequest {
  id: string;
  status: string;
  remarks: string;
}

export interface SalesOrderDropdownDto {
  id: string;
  salesOrderNo: string;
}

File: backend/Model/SalesOrder.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace YLWorks.Model
{
    public class SalesOrder : BaseEntity
    {
        public Guid Id { get; set; }
        public string SalesOrderNo { get; set; } = string.Empty;

        public Guid ClientId { get; set; }
        public Company Client { get; set; } = null!;

        public Guid? CompanyId {  get; set; }
        public Company? Company { get; set; }

        public Guid? QuotationId { get; set; }
        public Quotation? Quotation { get; set; }

        public Guid? ProjectId { get; set; }
        public Project? Project { get; set; }

        public string? Terms {  get; set; }

        public DateTime SODate { get; set; }
        public string Status { get; set; } = "Draft"; //Draft, Confirmed, InProgress, PartiallyDelivered, Delivered, Completed, Cancelled

        public string? ClientPONumber { get; set; }
        public DateTime? ClientPODate { get; set; }
        public string? ClientPOAttachment { get; set; }

        public decimal TotalAmount {  get; set; }
        public string? Notes { get; set; }
        public string? Remarks { get; set; }
        public ICollection<SalesOrderItem>? SalesOrderItems { get; set; } = new List<SalesOrderItem>();
        public ICollection<SalesOrderStatusHistory> SalesOrderStatusHistories { get; set; }
    = new List<SalesOrderStatusHistory>();
        public ICollection<DeliveryOrder> DeliveryOrders { get; set; }
    = new List<DeliveryOrder>();
    }

    public class SalesOrderItem
    {
        public Guid Id { get; set; }
        public Guid SalesOrderId { get; set; }

        public Guid? ParentId { get; set; }  
        public int SortOrder { get; set; }

        public string Type { get; set; }
        public bool IsGroup { get; set; }

        public string? Item {  get; set; }
        public string? Description { get; set; }

        public decimal? Quantity { get; set; }
        public decimal QuantityDelivered { get; set; }
        public decimal QuantityRemaining { get; set; }

        public string Unit { get; set; }
        public decimal? UnitPrice { get; set; }
        public decimal? TotalPrice { get; set; }
        public decimal? Discount { get; set; }

        public List<SalesOrderItem> Children { get; set; } = new();
    }

    public class SOItemBase
    {
        public Guid? Id { get; set; }
        public int SortOrder { get; set; }
        public string Type { get; set; }
        public bool IsGroup { get; set; }
        public Guid? ParentId { get; set; }
        public string? Item { get; set; }
        public string? Description { get; set; } = string.Empty;
        public decimal? Quantity { get; set; }
        public string Unit { get; set; } = "Unit";
        public decimal? UnitPrice { get; set; }
        public decimal DeliveredQuantity {  get; set; }
        public decimal? Discount { get; set; }
        public decimal? TotalPrice { get; set; }
        public List<SOItemRequest> Children { get; set; } = new();
    }

    public class SOItemRequest : SOItemBase { }

    public class CreateSalesOrderRequest
    {
        public string SalesOrderNo { get; set; } = string.Empty;

        public Guid? CompanyId { get; set; }
        public Guid ClientId { get; set; }
        public Guid? ProjectId { get; set; }
        public Guid? QuotationId { get; set; }

        public DateTime SODate { get; set; }
        public string? Terms { get; set; }

        public decimal TotalAmount { get; set; }

        public string? Notes { get; set; }
        public string? Remarks { get; set; }

        public string? ClientPONumber { get; set; }
        public DateTime? ClientPODate { get; set; }
        public IFormFile? ClientPOAttachment { get; set; }

        public List<SOItemBase>? SalesOrderItems { get; set; } = new();
    }

    public class UpdateSalesOrderRequest
    {
        public Guid Id { get; set; }

        public string SalesOrderNo { get; set; } = string.Empty;

        public Guid? CompanyId { get; set; }
        public Guid ClientId { get; set; }
        public Guid? ProjectId { get; set; }
        public Guid? QuotationId { get; set; }

        public DateTime SODate { get; set; }
        public string? Terms { get; set; }

        public decimal TotalAmount { get; set; }

        public string? Notes { get; set; }
        public string? Remarks { get; set; }

        public string? ClientPONumber { get; set; }
        public DateTime? ClientPODate { get; set; }
        public IFormFile? ClientPOAttachment { get; set; }

        public List<SOItemBase>? SalesOrderItems { get; set; } = new();
    }

    public class SalesOrderStatusHistory
    {
        public Guid Id { get; set; }

        public Guid SalesOrderId { get; set; }
        [JsonIgnore]
        public SalesOrder SalesOrder { get; set; } = null!;

        public string Status { get; set; } = string.Empty;

        public DateTime ActionAt { get; set; } = DateTime.UtcNow;

        public Guid? ActionUserId { get; set; }
        public User? ActionUser { get; set; }

        public string? Remarks { get; set; }
    }

    public class UpdateSalesOrderStatusRequest
    {
        public Guid Id { get; set; }
        public string Status { get; set; } = string.Empty;
        public string? Remarks { get; set; }
    }

    public class SalesOrderDropdownDto
    {
        public Guid Id { get; set; }
        public string SalesOrderNo { get; set; }
    }

}